### PR TITLE
Tweak ResponseOutputStream

### PR DESCRIPTION
**Pull request by hc4**
I'm using http server in single thread (through SynchonizationContext).
And getting deadlock on ResponseOutputStream disposing, because it tries to synchoniously flush written data and can't return to execution thread, as it is blocked by sync wait.

Also there was a bug in buffered write logic, In case of written data size >MAX_RESPONSE_BUFFER the chunk of data caused overflow has been written twice.

### CODE REVIEW

File: Ceen.Httpd/ResponseOutputStream.cs [hunk 1]
```diff
@@ -155,23 +158,23 @@ public override async Task WriteAsync(byte[] buffer, int offset, int count, Canc
...
 					m_passThrough = true;
+				}
 				else
 				{
 					if (m_buffer == null)
 						m_buffer = new MemoryStream();
 
 					await m_buffer.WriteAsync(buffer, offset, count, cancellationToken);
 					m_written += count;
-
```
Comment (kenkendk) [inline Ceen.Httpd/ResponseOutputStream.cs]: This fix is certainly correct. The data has been added to the buffer, so the code cannot continue. An alternative would be to set `count = 0;` so line 176 does not write the data again.

Comment (hc4) [reply Ceen.Httpd/ResponseOutputStream.cs]: count=0 won't prevent SetLengthAndFlushAsync, which will break buffering logic.
Also it looks like a hack :)

Comment (kenkendk) [reply Ceen.Httpd/ResponseOutputStream.cs]: It _should_ flush if the maximum amount of buffer space has been used. But I am good with returning.

**Comment by kenkendk** [comment]
For the synchronous deadlock, can you post a minimal example that deadlocks, so I can try to figure out another way than simply starting a new task?

File: Ceen.Httpd/ResponseOutputStream.cs (hunk 1 above)
Comment (hc4) [reply Ceen.Httpd/ResponseOutputStream.cs]: If there won't be enough space in buffer, codition on line 161 will be met and buffer will be flushed.

**Comment by hc4** [comment]
> For the synchronous deadlock, can you post a minimal example that deadlocks
Just try to write and dispose ResponseOutputStream from UI thread (Forms or WPF).

As I wrote - best solution is to add sync version of writing logic.
But doesn't looks easy...

**Comment by hc4** [comment]
> another way than simply starting a new task

I'm afraid it is ompossible.
Current version will spawn new task on first await, because main task is blocked by sync Wait.

**Comment by kenkendk** [comment]
I like the explicit solution with a waiter queue.

**Pull request merged**